source/d3d9/runtime_d3d9.cpp

```diff
@@ -1394,4 +1767,164 @@ namespace reshade::d3d9
+
+		if (!SUCCEEDED(hr))
+		{
+			LOG(ERROR) << "Failed to create depth replacement texture! HRESULT is '" << std::hex << hr << std::dec << "'.";
+
+			return nullptr;
+		}
+
+		return depthstencil_replacement;
+	}
+
+	// retrieve the depth buffer surface replacement ref, depending on the detection mode
+	com_ptr<IDirect3DSurface9> runtime_d3d9::get_depthstencil_replacement()
+	{
+		const auto it = _depth_clearing_table.find(_clear_idx);
+
+		if (it != _depth_clearing_table.end())
+		{
+			return it->second.depthstencil;
+		}
+
+		return _depthstencil_replacement;
+	}
+
+	// retrieve the depth buffer texture replacement ref, depending on the detection mode
+	com_ptr<IDirect3DTexture9> runtime_d3d9::get_depthstencil_texture()
+	{
+		const auto it = _depth_clearing_table.find(_clear_idx);
+
+		if (it != _depth_clearing_table.end())
+		{
+			return it->second.depthtexture;
+		}
+		return _depthstencil_texture;
+	}
+
+	// retrieve the best depth buffer clearing instance, depending on the options selected
+	int runtime_d3d9::get_best_preserve_starting_index(bool multi)
+	{
+		int result = multi ? -1 : 0;
+
+		for (const auto &it : _depth_clearing_table)
+		{
+			bool best = (it.second.vertices_count >= _db_vertices);
+			if (_auto_preserve_on_drawcalls)
+				best = (it.second.drawcall_count >= _db_drawcalls);
+
+			if (best)
+			{
+				_db_vertices = it.second.vertices_count;
+				_db_drawcalls = it.second.drawcall_count;
+			}
+
+			// only when multiple depth buffer replacement textures
+			if (best && multi)
+			{
+				if (_auto_preserve_on_vertices)
+				{
+					// compare with the max number of vertices from the previous frame. If too much difference, exclude this candidate
+					if (it.second.vertices_count < (_previous_best_vertices / 10))
+						best = false;
+				}
+				if (_auto_preserve_on_drawcalls)
+				{
+					// compare with the max number of drawcalls from the previous frame. If too much difference, exclude this candidate
+					if (it.second.drawcall_count < (_previous_best_drawcalls / 10))
+						best = false;
+				}
+			}
+
+			if (best)
+				result = it.first;
+		}
+
+		return result;
+	}
+
+	bool runtime_d3d9::check_depthstencil_size(com_ptr<IDirect3DSurface9> depthstencil)
+	{
+		D3DSURFACE_DESC desc;
+		depthstencil->GetDesc(&desc);
+
+		// check the size of the current depth buffer
+		if (!_disable_depth_buffer_size_restriction)
+		{
+			if ((desc.Width < _width * 0.95 || desc.Width > _width * 1.05) ||
+				(desc.Height < _height * 0.95 || desc.Height > _height * 1.05))
+			{
+				return false;
+			}
+		}
+		else
+		{
+			// allow to retrieve depth buffers having greater dimensions than the viewport's ones (fix depth buffer detection in games like Vanquish)
+			if (desc.Width < _width * 0.95 ||
+				desc.Height < _height * 0.95)
+			{
+				return false;
+			}
+		}
+
+		return true;
+	}
+
+	// preservation decision
+	bool runtime_d3d9::is_preserved()
```

Comment:
I'd inline this and use something like
```
if (_multi_depthstencil ?
    // comment
    _clear_idx <= _preserve_starting_index :
    // comment
    _clear_idx >= _preserve_starting_index)
{
}
```
Together with above suggestion of moving `before_clear` and `after_clear` into a single function, this would only be called once anyway.